File: src/main/java/toast/client/modules/config/Setting.java
package toast.client.modules.config;

import com.google.gson.annotations.SerializedName;
import toast.client.utils.Config;

public class Setting {
    @SerializedName("Current Mode")
    private String mode;
    @SerializedName("Enabled")
    private Boolean enabled;
    @SerializedName("Current Value")
    private Double value;

    public Setting(String mode) {
        setMode(mode);
    }

    public Setting(Double value) {
        setValue(value);
    }

    public Setting(Boolean enabled) {
        setEnabled(enabled);
    }

    public String getMode() {
        return this.mode;
    }

    public void setMode(String newMode) {
        this.mode = newMode;
        this.value = null;
        this.enabled = null;
        Config.writeConfig();
    }

    public double getValue() {
        return this.value;
    }

    public void setValue(double newValue) {
        this.mode = null;
        this.value = newValue;
        this.enabled = null;
        Config.writeConfig();
    }

    public boolean isEnabled() {
        return this.enabled;
    }

    public void setEnabled(boolean enabled) {
        this.mode = null;
        this.value = null;
        this.enabled = enabled;
        Config.writeConfig();
    }

    public int getType() {
        if (mode != null) return 0;
        else if (value != null) return 1;
        else if (enabled != null) return 2;
        else return 3;
    }
}

File: src/main/java/toast/client/modules/config/Setting.kt
package toast.client.modules.config

import com.google.gson.annotations.SerializedName
import toast.client.ToastClient

class Setting {
    @SerializedName("Current Mode")
    private var mode: String? = null

    @SerializedName("Enabled")
    private var enabled: Boolean? = null

    @SerializedName("Current Value")
    private var value: Double? = null

    constructor(mode: String) {
        setMode(mode)
    }

    constructor(value: Double) {
        setValue(value)
    }

    constructor(enabled: Boolean) {
        setEnabled(enabled)
    }

    fun getMode(): String? {
        return mode
    }

    fun setMode(newMode: String) {
        mode = newMode
        value = null
        enabled = null
        ToastClient.CONFIG_MANAGER.writeConfig()
    }

    fun getValue(): Double = value!!

    fun setValue(newValue: Double) {
        mode = null
        value = newValue
        enabled = null
        ToastClient.CONFIG_MANAGER.writeConfig()
    }

    fun isEnabled(): Boolean = enabled!!

    fun setEnabled(enabled: Boolean) {
        mode = null
        value = null
        this.enabled = enabled
        ToastClient.CONFIG_MANAGER.writeConfig()
    }

    val type: Int
        get() = if (mode != null) 0 else if (value != null) 1 else if (enabled != null) 2 else 3
}